Mathlib/Combinatorics/Optimization/ValuedCSP.lean

```diff
@@ -0,0 +1,216 @@
+Authors: Martin Dvorak
+-/
+import Mathlib.Algebra.Order.Monoid.Defs
+import Mathlib.Data.Finset.Basic
+import Mathlib.Data.Fin.VecNotation
+import Mathlib.Data.Rat.Order -- will be trimmed after deleting examples
+import Mathlib.Tactic.Linarith
+import Mathlib.Tactic.Cases
+import Mathlib.Data.Bool.Basic
+
+/-!
+
+# General-Valued Constraint Satisfaction Problems
+
+General-Valued CSP is a very broad class of problems in discrete optimization.
+General-Valued CSP subsumes Min-Cost-Hom (including 3-SAT for example) and Finite-Valued CSP.
+
+## Main definitions
+* `ValuedCspTemplate`: A VCSP template; fixes a domain, a codomain, and allowed cost functions.
+* `ValuedCspTerm`: One summand in a VCSP instance; calls a concrete function from given template.
+* `ValuedCspTerm.evalSolution`: An evaluation of the VCSP term for given solution.
+* `ValuedCspInstance`: An instance of a VCSP problem over given template.
+* `ValuedCspInstance.evalSolution`: An evaluation of the VCSP instance for given solution.
+* `ValuedCspInstance.optimumSolution`: Is given solution a minimum of the VCSP instance?
+
+## References
+* [D. A. Cohen, M. C. Cooper, P. Creed, P. G. Jeavons, S. Živný,
+   *An Algebraic Theory of Complexity for Discrete Optimisation*][cohen2012]
+
+-/
+
+def n1ary_of_unary {α β : Type _} (f : α → β) : (Fin 1 → α) → β :=
+  fun a => f (a 0)
+
+def n2ary_of_binary {α β : Type _} (f : α → α → β) : (Fin 2 → α) → β :=
+  fun a => f (a 0) (a 1)
+
+/-- A template for a valued CSP problem with costs in `C`. -/
+structure ValuedCspTemplate (C : Type _) [LinearOrderedAddCommMonoid C] where
+  /-- Domain of "labels" -/
+  D : Type
+  /-- Cost functions from `D^k` to `C` for any `k` -/
+  F : Set (Σ (k : ℕ), (Fin k → D) → C)
+
+variable {C : Type _} [LinearOrderedAddCommMonoid C]
+
+/-- A term in a valued CSP instance over the template `Γ`. -/
+structure ValuedCspTerm (Γ : ValuedCspTemplate C) (ι : Type _) where
+  /-- Arity of the function -/
+  k : ℕ
+  /-- Which cost function is instantiated -/
+  f : (Fin k → Γ.D) → C
+  /-- The cost function comes from the template -/
+  inΓ : ⟨k, f⟩ ∈ Γ.F
+  /-- Which variables are plugged as arguments to the cost function -/
+  app : Fin k → ι
+
+def valuedCspTerm_of_unary {Γ : ValuedCspTemplate C} {ι : Type _} {f₁ : Γ.D → C}
+    (ok : ⟨1, n1ary_of_unary f₁⟩ ∈ Γ.F) (i : ι) : ValuedCspTerm Γ ι :=
+  ⟨1, n1ary_of_unary f₁, ok, ![i]⟩
+
+def valuedCspTerm_of_binary {Γ : ValuedCspTemplate C} {ι : Type _} {f₂ : Γ.D → Γ.D → C}
+    (ok : ⟨2, n2ary_of_binary f₂⟩ ∈ Γ.F) (i j : ι) : ValuedCspTerm Γ ι :=
+  ⟨2, n2ary_of_binary f₂, ok, ![i, j]⟩
+
+/-- Evaluation of a `Γ` term `t` for given solution `x`. -/
+def ValuedCspTerm.evalSolution {Γ : ValuedCspTemplate C} {ι : Type _}
+    (t : ValuedCspTerm Γ ι) (x : ι → Γ.D) : C :=
+  t.f (x ∘ t.app)
+
+/-- A valued CSP instance over the template `Γ` with variables indexed by `ι`.-/
+def ValuedCspInstance (Γ : ValuedCspTemplate C) (ι : Type _) : Type :=
```

Comment:
Also not clear on the need for this if it's just a list of things already defined, and whether it should be named `ValuedCspInstance` if it's just a list of optimization problems. I suspect there is some naming convection in the literature that differs from the Mathlib meaning of instance.

Reply:
That's not really a list of optimization problems. `ValuedCspInstance` is one optimization problem; it consists of terms that overlap in the variables referred from them. If you want to optimize a function given some constraints, you provide the function to be minimized as a `ValuedCspTerm` with rational values and the constraints as `ValuedCspTerm` with {0, infinity} values.

Note that I am not referring to the meaning of a typeclass instance (which would be a Lean terminology, not an optimization terminology).

Reply:
For example, we want to minimize 1/x + 1/y on { (x, y) | x > 0, y > 0, x + y = 2 }.
The domain will be positive rationals.
The template will contain (at least) the following two functions:
* a unary function that calculates the reciprocal
* a binary function that returns zero when the arguments sum up to 2 and infinity otherwise
The instance then consists of two terms that use the unary function and one term that uses the binary function.
The optimum solution is (1, 1), represented as `fun (_ : Fin 2) => (1 : ℚ)` in Lean.

Reply:
I see. This is too foreign to me :(
I'm used to studying optimization theory in the convex programming context which takes a very different approach. I don't think I'll be able to provide any meaningful feedback here, sorry.